services/databaseEncryptionService.ts

```diff
@@ -0,0 +1,348 @@
+import * as SecureStore from 'expo-secure-store';
+import * as Crypto from 'expo-crypto';
+import { gcm } from '@noble/ciphers/aes.js';
+
+const SECURE_STORE_KEYS = {
+  ENCRYPTED_DEK: 'encrypted_dek',
+  KEK: 'kek',
+  ENCRYPTION_MODE: 'encryption_mode',
+};
+
+export type EncryptionMode = 'basic' | 'protected';
+
+export class EncryptionError extends Error {
+  constructor(public code: string, message: string) {
+    super(message);
+    this.name = 'EncryptionError';
+  }
+}
+
+export const ERROR_CODES = {
+  USER_CANCELLED: 'USER_CANCELLED',
+  KEY_CORRUPTION: 'KEY_CORRUPTION',
+  SECURE_STORE_UNAVAILABLE: 'SECURE_STORE_UNAVAILABLE',
+  REWRAP_FAILED: 'REWRAP_FAILED',
+  AUTH_IN_PROGRESS: 'AUTH_IN_PROGRESS',
+} as const;
+
+let dekCache: string | null = null;
+
+function generateRandomKey(): string {
+  const randomBytes = Crypto.getRandomBytes(32);
+  return Array.from(randomBytes)
+    .map(b => b.toString(16).padStart(2, '0'))
```

Comment:
The encoding in this file is needlessly complicated. You have two routines that convert hex to bytes, and on top of that you're also doing base64 encoding. You could just use base64 to rule them all.